File: examples/0_basic/3_deactivate_did.rs
// Copyright 2020-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use examples::create_did_document;
use examples::get_funded_client;
use examples::get_memstorage;
use examples::TEST_GAS_BUDGET;
use identity_iota::iota::IotaDID;
use identity_iota::iota::IotaDocument;

/// Demonstrates how to deactivate a DID in an identity.
#[tokio::main]
async fn main() -> anyhow::Result<()> {
  // create new client to interact with chain and get funded account with keys
  let storage = get_memstorage()?;
  let identity_client = get_funded_client(&storage).await?;

  // create new DID document and publish it
  let (document, _) = create_did_document(&identity_client, &storage).await?;

  println!("Published DID document: {document:#}");

  let did: IotaDID = document.id().clone();

  // Deactivate the DID by publishing an empty document.
  // This process can be reversed since the identity is not destroyed.
  // Deactivation may only be performed by a controller of the identity.
  identity_client.deactivate_did_output(&did, TEST_GAS_BUDGET).await?;

  // Resolving a deactivated DID returns an empty DID document
  // with its `deactivated` metadata field set to `true`.
  let deactivated: IotaDocument = identity_client.resolve_did(&did).await?;
  println!("Deactivated DID document: {deactivated:#}");
  assert_eq!(deactivated.metadata.deactivated, Some(true));

  // Re-activate the DID by publishing a valid DID document.
  let reactivated: IotaDocument = identity_client
    .publish_did_document_update(document.clone(), TEST_GAS_BUDGET)
    .await?;
  println!("Reactivated DID document result: {reactivated:#}");

  let resolved: IotaDocument = identity_client.resolve_did(&did).await?;
  println!("Reactivated DID document resolved from chain: {resolved:#}");

  Ok(())
}

File: bindings/wasm/identity_wasm/examples/src/0_basic/3_deactivate_did.ts
// Copyright 2020-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { IotaClient } from "@iota/iota-sdk/client";
import { createDocumentForNetwork, getFundedClient, getMemstorage, NETWORK_URL, TEST_GAS_BUDGET } from "../util";

/** Demonstrates how to deactivate a DID of an identity. */
export async function deactivateIdentity() {
    // create new clients and create new account
    const iotaClient = new IotaClient({ url: NETWORK_URL });
    const network = await iotaClient.getChainIdentifier();
    const storage = getMemstorage();
    const identityClient = await getFundedClient(storage);
    const [unpublished] = await createDocumentForNetwork(storage, network);

    // create new identity for this account and publish document for it
    const { output: identity } = await identityClient
        .createIdentity(unpublished)
        .finish()
        .buildAndExecute(identityClient);
    const did = identity.didDocument().id();

    // Resolve the latest state of the document.
    // Technically this is equivalent to the document above.
    const resolved = await identityClient.resolveDid(did);
    console.log("Resolved DID document:", JSON.stringify(resolved, null, 2));

    const controllerToken = await identity.getControllerToken(identityClient);
    // Deactivate the DID.
    await identity
        .deactivateDid(controllerToken!)
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(identityClient);

    // Resolving a deactivated DID returns an empty DID document
    // with its `deactivated` metadata field set to `true`.
    let deactivated = await identityClient.resolveDid(did);
    console.log("Deactivated DID document:", JSON.stringify(deactivated, null, 2));
    if (deactivated.metadataDeactivated() !== true) {
        throw new Error("Failed to deactivate DID document");
    }

    // Re-activate the DID by publishing a valid DID document.
    console.log("Publishing this:", JSON.stringify(resolved, null, 2));
    await identity
        .updateDidDocument(resolved, controllerToken!)
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(identityClient);

    // Resolve the reactivated DID document.
    let resolvedReactivated = await identityClient.resolveDid(did);
    console.log("Reactivated DID document:", JSON.stringify(resolvedReactivated, null, 2));
    if (resolvedReactivated.metadataDeactivated() === true) {
        throw new Error("Failed to reactivate DID document");
    }
}
